src/main/java/com/farao_community/farao/gridcapa/gateway/filters/TokenValidatorGatewayFilterFactory.java

```diff
@@ -139,31 +140,56 @@ private Mono<Void> handleTokenAsJwt(String token, ServerWebExchange exchange, Ga
     }
 
     private Mono<Void> validateTokenAndSetHeaderUserId(FilterInfos filterInfos) {
-        boolean cacheRefreshed = false;
-        if (jwkSetCache == null) {
-            try {
-                addJwkSetToCache();
-                cacheRefreshed = true;
-            } catch (ParseException e) {
-                LOGGER.info(PARSING_ERROR, filterInfos.exchange().getRequest().getPath());
-                return completeWithCode(filterInfos.exchange(), HttpStatus.INTERNAL_SERVER_ERROR);
-            }
+        final Mono<Void> cacheMono;
+
+        final boolean cacheWasNull = jwkSetCache == null;
+        if (cacheWasNull) {
+            cacheMono = addJwkSetToCache();
+        } else {
+            cacheMono = Mono.empty();
         }
 
+        return cacheMono
+                .then(Mono.defer(() -> filterWithValidatedJwt(filterInfos, cacheWasNull)));
+    }
+
+    private Mono<Void> filterWithValidatedJwt(final FilterInfos filterInfos,
+                                              final boolean cacheWasNull) {
         try {
             validateJwt(filterInfos);
         } catch (JOSEException | BadJOSEException e) {
-            return handleValidationException(filterInfos, cacheRefreshed, e);
+            return handleValidationException(filterInfos, cacheWasNull, e);
         }
+        final ServerWebExchange newExchange = setHeaderUserIdInNewExchange(
+                filterInfos.exchange(),
+                filterInfos.jwtClaimsSet().getSubject()
+        );
+        return filterInfos.chain().filter(newExchange);
+    }
 
-        setHeaderUserId(filterInfos.exchange(), filterInfos.jwtClaimsSet().getSubject());
-
-        return filterInfos.chain().filter(filterInfos.exchange());
+    private Mono<Void> addJwkSetToCache() {
+        return webClient.get()
+                .uri(jwkSetUri)
+                .retrieve()
+                .bodyToMono(String.class)
+                .<JWKSet>handle((jwkSetStr, sink) -> {
+                    try {
+                        sink.next(parseJwkSet(jwkSetStr));
+                    } catch (RuntimeException | ParseException e) {
+                        LOGGER.error(PARSING_ERROR, e);
+                        sink.error(new JwkSetParsingException(PARSING_ERROR, e));
+                    }
+                })
+                .doOnNext(jwkSet -> jwkSetCache = jwkSet)
+                .then()
+                .onErrorResume(e -> {
+                    LOGGER.error("Impossible to update JWKSet cache : {}", e.getMessage(), e);
+                    return Mono.error(e);
+                });
     }
 
-    private void addJwkSetToCache() throws ParseException {
-        String jwkSetString = restTemplate.getForObject(jwkSetUri, String.class);
-        jwkSetCache = JWKSet.parse(jwkSetString);
+    private JWKSet parseJwkSet(String jwkSetString) throws ParseException {
```

Comment:
Not sure about the usefulness of this method with single action anymore. Inline it?